syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt

```diff
@@ -83,49 +90,50 @@ class BlockPuller internal constructor(private val connectionHandler: Connection
             }
 
             override fun close() {
-                missingHashes.clear()
-                hashList.clear()
                 blocksByHash.clear()
             }
         }
-        synchronized(lock) {
-            hashList.addAll(fileBlocks.blocks.map { it.hash })
-            missingHashes.addAll(hashList)
-            for (block in fileBlocks.blocks) {
-                if (missingHashes.contains(block.hash)) {
-                    val requestId = Math.abs(Random().nextInt())
-                    requestIds.add(requestId)
-                    connectionHandler.sendMessage(Request.newBuilder()
-                            .setId(requestId)
-                            .setFolder(fileBlocks.folder)
-                            .setName(fileBlocks.path)
-                            .setOffset(block.offset)
-                            .setSize(block.size)
-                            .setHash(ByteString.copyFrom(Hex.decode(block.hash)))
-                            .build())
-                    logger.debug("sent request for block, hash = {}", block.hash)
+
+        for (block in fileBlocks.blocks.distinctBy { it.hash }) {
+            val requestId = responseHandler.registerListener {
+                response ->
+
+                try {
+                    NetworkUtils.assertProtocol(response.code == ErrorCode.NO_ERROR, { "received error response, code = ${response.code}" })
+                    val data = response.data.toByteArray()
+                    val hash = Hex.toHexString(MessageDigest.getInstance("SHA-256").digest(data))
+
+                    if (hash != block.hash) {
+                        throw IllegalStateException("expected block with hash ${block.hash}, but got block with hash $hash")
+                    }
```

Comment:
I think throwing here is not a good idea, unless you only support downloading from a single peer. Otherwsise i'd try a different peer for the same block.

Reply:
Every block is only requested once currently. Retrying (from other peers or the same one at a later time) would be useful, but throwing is the best solution to indicate an problem (I would later add an timeout and throw an exception on a timeout too).